packages/opentelemetry-types/src/trace/instrumentation/Plugin.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Tracer } from '../tracer';
+
+/** Interface Plugin to apply patch. */
+export interface Plugin {
+  /**
+   * Method that enables the instrumentation patch.
+   * @param moduleExports The value of the `module.exports` property that would
+   *     normally be exposed by the required module. ex: `http`, `https` etc.
+   * @param tracer a tracer instance
+   */
+  enable<T>(moduleExports: T, tracer: Tracer): T;
```

Comment:
Since all plugins will be called in the same way by some sort of instrumenter, should there be a method/parameter to configure the plugin? Having that in the constructor would mean we can't validate it from the interface.

Reply:
Sure. Can we add another optional param (`config`)? What are the possible attributes for `config`? ex: version, default attributes like these?

Reply:
I think config is basically just an object and the possible properties depend on each plugin.

Reply:
One other thing that seems to be missing too is a way to define the supported versions/modules on every plugin in a standard way.

Reply:
```
enable(moduleExports: T, tracer: Tracer, 
       version: string, config?: {[key: unknown]: unknown}): T;
```
WDYT?

Reply:
> config?: {[key: unknown]: unknown}

I think it's safe to assume keys will be strings, especially if it's an object since other types are not possible.

> version: string

I think maybe the module name and version filters should be protected properties on the base tracer that the implementations can set. I don't think these should be set by the caller.

Reply:
Added the `config` object. ptal